File: packages/protocol/src/calls.ts
/**
 * FR-15 — Call control-plane wire contract.
 *
 * The call *records* (CallRoom / CallInvite / CallParticipant) sync to clients
 * as ordinary object records, the call event log rides the `CallEventStream`,
 * and WebRTC SDP/ICE signaling rides the existing `SignalSend`/`SignalDeliver`
 * frames with the `WebRTCSignal` type — none of that needs a new frame.
 *
 * What *does* need a new frame is the set of server-authoritative lifecycle
 * *commands*: create / join / leave / accept / set-media-state / end. These are
 * not plain object upserts — the server validates every transition, allocates
 * media sessions, mints per-participant join grants, and emits durable events.
 * So this module adds a request/response RPC frame pair that mirrors the
 * Append/ObjectUpsert request shape (a client-supplied `requestId` for
 * correlation) but returns a typed result the plain `Ack` frame can't carry
 * (the room snapshot + the participant + the media join grant).
 *
 * These shapes are the canonical wire contract the TS, Swift, and Android
 * clients all mirror. They intentionally line up with the server's
 * `CallControlPlane` record/result types (`apps/server/src/calls`).
 */

import type { PlainObject } from "./schema.js";

/** Media topology a call's media session is brokered over (FR-78). */
export type CallTransport = "p2p" | "sfu";

/** Audio-only vs audio+video call (FR-79). */
export type CallKind = "audio" | "video";

/** Lifecycle states a `CallRoom` moves through. */
export type CallRoomState = "ringing" | "active" | "ended";

/** Lifecycle states a `CallInvite` moves through. */
export type CallInviteState = "ringing" | "accepted" | "declined" | "cancelled";

/** Lifecycle states a `CallParticipant` moves through. */
export type CallParticipantState = "joined" | "left";

/**
 * Kinds of WebRTC signal that ride the `WebRTCSignal` relay (opaque payload).
 *
 * `"keyEpoch"` (FR-156) carries an E2EE key-epoch announcement for per-room
 * call encryption: when membership changes, the rotation initiator wraps the
 * fresh epoch key and announces it over this same relay (the control plane
 * forwards the opaque `payload` byte-for-byte, exactly as it does `"sfuToken"`).
 * It is additive — a peer that doesn't understand it simply ignores it.
 */
export type WebRTCSignalKind =
  | "offer"
  | "answer"
  | "ice"
  | "renegotiate"
  | "sfuToken"
  | "keyEpoch";

/**
 * Server-authoritative call room record. Mirrors the server's
 * `CallRoomRecord`. Synced to clients as a `CallRoom` object; also returned
 * inline by create/join/leave/end commands so the caller gets the post-command
 * snapshot without waiting for the object delta.
 */
export interface CallRoomRecord {
  readonly id: string;
  readonly conversationId: string;
  readonly state: CallRoomState;
  readonly createdBy: string;
  readonly kind: CallKind;
  readonly createdAt: string;
  readonly startedAt?: string;
  readonly endedAt?: string;
  readonly mediaSessionId?: string;
  readonly transport?: string;
}

/** Server-authoritative invite record. Mirrors the server's `CallInviteRecord`. */
export interface CallInviteRecord {
  readonly id: string;
  readonly callId: string;
  readonly inviteeUserId: string;
  readonly status: CallInviteState;
  readonly invitedBy: string;
  readonly invitedAt: string;
  readonly respondedAt?: string;
}

/**
 * Server-authoritative participant record. Mirrors the server's
 * `CallParticipantRecord`. Carries the per-participant call-presence surface
 * (FR-82): mic / camera / screen-share. `speaking` and `networkQuality` are
 * optional, client-derived presence fields that ride the same record on the
 * wire when a media adapter reports them.
 */
export interface CallParticipantRecord {
  readonly id: string;
  readonly callId: string;
  readonly userId: string;
  readonly deviceId: string;
  readonly state: CallParticipantState;
  readonly joinedAt: string;
  readonly leftAt?: string;
  readonly micEnabled: boolean;
  readonly cameraEnabled: boolean;
  readonly screenSharing: boolean;
  /** FR-82: whether the participant is currently detected as speaking. */
  readonly speaking?: boolean;
  /** FR-82: coarse network-quality bucket for this participant's media link. */
  readonly networkQuality?: CallNetworkQuality;
}

/** Coarse network-quality bucket surfaced on the participant model (FR-82). */
export type CallNetworkQuality = "unknown" | "poor" | "fair" | "good" | "excellent";

/**
 * A short-lived, participant-scoped credential the client presents to the
 * media plane to join the actual media session. Mirrors the server's
 * `MediaJoinGrant` (FR-78). `connection` is opaque adapter metadata (ICE/TURN
 * for p2p, signaling URL for SFU).
 */
export interface CallMediaGrant {
  readonly callId: string;
  readonly mediaSessionId: string;
  readonly userId: string;
  readonly deviceId: string;
  readonly token: string;
  readonly expiresAt: string;
  readonly connection?: Readonly<Record<string, string>>;
}

/** Partial media-state mutation for the `setMediaState` command (FR-82). */
export interface CallMediaStatePatch {
  readonly micEnabled?: boolean;
  readonly cameraEnabled?: boolean;
  readonly screenSharing?: boolean;
}

// -- command request bodies -------------------------------------------------

export interface CreateCallCommand {
  readonly op: "create";
  readonly conversationId: string;
  /** Users to invite. Must be non-empty and must exclude the caller. */
  readonly inviteeUserIds: readonly string[];
  readonly kind?: CallKind;
  /** Region hint forwarded to the media plane (honored if supported). */
  readonly regionHint?: string;
}

export interface JoinCallCommand {
  readonly op: "join";
  readonly callId: string;
}

export interface AcceptCallCommand {
  readonly op: "accept";
  readonly callId: string;
}

export interface LeaveCallCommand {
  readonly op: "leave";
  readonly callId: string;
}

export interface EndCallCommand {
  readonly op: "end";
  readonly callId: string;
}

export interface SetCallMediaStateCommand {
  readonly op: "setMediaState";
  readonly callId: string;
  readonly media: CallMediaStatePatch;
}

// -- SFU media-negotiation commands (FR-155) --------------------------------
//
// The SFU media plane (FR-83) forwards real-time media through a server-side
// mediasoup router. After a client `join`s and receives its grant (router RTP
// capabilities + send/recv transport params), it negotiates media by driving
// these server-authoritative ops over the SAME `CallCommand` frame: it loads a
// mediasoup-client `Device`, builds its transports, and then completes the DTLS
// handshake (`connectTransport`), starts producing its own tracks (`produce`),
// and consumes other participants' producers (`consume`). These are no-ops on a
// P2P media plane — a P2P server Nacks them. All structured params
// (dtlsParameters, rtpParameters, rtpCapabilities) are carried as opaque
// JSON-serializable objects, mirroring the server's `SfuBackend` shapes.

/** Opaque, JSON-serializable DTLS parameters from a mediasoup-client transport. */
export type CallSfuDtlsParameters = Record<string, unknown>;

/** Opaque, JSON-serializable RTP parameters for a produced/consumed track. */
export type CallSfuRtpParameters = Record<string, unknown>;

/** Opaque, JSON-serializable RTP capabilities of the consuming client's Device. */
export type CallSfuRtpCapabilities = Record<string, unknown>;

/** Media kind a producer/consumer carries (mirrors the server's `MediaKind`). */
export type CallSfuMediaKind = "audio" | "video";

/**
 * Complete the DTLS handshake for one of this participant's transports. Sent by
 * the client when a mediasoup-client transport fires its `connect` event.
 */
export interface ConnectSfuTransportCommand {
  readonly op: "sfuConnectTransport";
  readonly callId: string;
  /**
   * The join nonce the client received in its {@link CallMediaGrant} (FR-78).
   * The server re-derives and verifies it (signature + expiry + identity
   * binding) before forwarding the op to the SFU. Additive: older servers
   * ignore it, newer servers require it for SFU media ops.
   */
  readonly token: string;
  readonly transportId: string;
  readonly dtlsParameters: CallSfuDtlsParameters;
}

/**
 * Start producing one of this participant's tracks on its send transport. Sent
 * when a mediasoup-client send transport fires its `produce` event. The result
 * carries the server-assigned `producerId` the client returns to the transport.
 */
export interface ProduceSfuCommand {
  readonly op: "sfuProduce";
  readonly callId: string;
  /** Join nonce re-verified server-side before producing (see {@link ConnectSfuTransportCommand.token}). */
  readonly token: string;
  readonly transportId: string;
  readonly kind: CallSfuMediaKind;
  readonly rtpParameters: CallSfuRtpParameters;
}

/**
 * Consume another participant's producer onto this participant's recv transport.
 * The result carries the consumer params the client feeds to
 * `recvTransport.consume(...)`.
 */
export interface ConsumeSfuCommand {
  readonly op: "sfuConsume";
  readonly callId: string;
  /** Join nonce re-verified server-side before consuming (see {@link ConnectSfuTransportCommand.token}). */
  readonly token: string;
  readonly transportId: string;
  readonly producerId: string;
  readonly rtpCapabilities: CallSfuRtpCapabilities;
}

/** Discriminated union of every call control-plane command. */
export type CallCommandOp =
  | CreateCallCommand
  | JoinCallCommand
  | AcceptCallCommand
  | LeaveCallCommand
  | EndCallCommand
  | SetCallMediaStateCommand
  | ConnectSfuTransportCommand
  | ProduceSfuCommand
  | ConsumeSfuCommand;

/** Names of the supported command operations. */
export type CallCommandName = CallCommandOp["op"];

/**
 * A call command frame body. Carries a client-supplied `requestId` for
 * correlation with the matching {@link CallCommandResultPayload}, exactly like
 * Append/ObjectUpsert correlate with Ack/Nack.
 */
export interface CallCommandPayload {
  readonly requestId: string;
  readonly command: CallCommandOp;
}

// -- command result body ----------------------------------------------------

/**
 * The server's reply to a {@link CallCommandPayload}. `requestId` echoes the
 * request. Each field is populated for the commands that produce it:
 *  - `create`        → `room`, `invites`
 *  - `join`          → `room`, `participant`, `mediaGrant`
 *  - `accept`        → `invite`
 *  - `leave` / `end`     → `room`
 *  - `setMediaState`     → `participant`
 *  - `sfuConnectTransport` → (no payload — success is the absence of a Nack)
 *  - `sfuProduce`        → `producer`
 *  - `sfuConsume`        → `consumer`
 *
 * Failures come back as the existing `Nack` frame keyed by the same
 * `requestId` (no separate error channel), so clients reuse their Ack/Nack
 * correlation plumbing.
 */
export interface CallCommandResultPayload {
  readonly requestId: string;
  readonly op: CallCommandName;
  readonly room?: CallRoomRecord;
  readonly invites?: readonly CallInviteRecord[];
  readonly participant?: CallParticipantRecord;
  readonly mediaGrant?: CallMediaGrant;
  readonly invite?: CallInviteRecord;
  /** `sfuProduce` → the server-assigned producer the client returns to the transport. */
  readonly producer?: CallSfuProduceResult;
  /** `sfuConsume` → the consumer params the client feeds to `recvTransport.consume`. */
  readonly consumer?: CallSfuConsumeResult;
}

/** Result of a {@link ProduceSfuCommand}: the new server-side producer id. */
export interface CallSfuProduceResult {
  readonly producerId: string;
  readonly kind: CallSfuMediaKind;
}

/** Result of a {@link ConsumeSfuCommand}: everything `recvTransport.consume` needs. */
export interface CallSfuConsumeResult {
  readonly consumerId: string;
  readonly producerId: string;
  readonly kind: CallSfuMediaKind;
  readonly rtpParameters: CallSfuRtpParameters;
}

/**
 * Build the opaque `WebRTCSignal` value carried by a `SignalSend` frame. The
 * server is a control plane only: it relays `payload` byte-for-byte without
 * interpreting it (it's the SDP / ICE candidate the media layer produces).
 */
export interface WebRTCSignalValue extends PlainObject {
  readonly senderDeviceId: string;
  readonly recipientDeviceId?: string;
  readonly kind: WebRTCSignalKind;
  readonly payload: Uint8Array;
}

/** Canonical signal type name for WebRTC relay (matches the server schema). */
export const WEBRTC_SIGNAL_TYPE = "WebRTCSignal";

File: crates/frick-protocol/src/calls.rs
//! FR-15 call control-plane wire contract (`packages/protocol/src/calls.ts`).
//!
//! Call records sync as ordinary objects and WebRTC signaling rides
//! `SignalSend`/`SignalDeliver`; this module is the request/response RPC
//! pair for server-authoritative lifecycle commands. Opaque mediasoup
//! parameter bags stay dynamic [`Value`] maps, forwarded byte-for-byte.

use indexmap::IndexMap;
use serde::{Deserialize, Serialize};

use crate::value::{Value, string_enum};

string_enum! {
    /// Media topology a call's media session is brokered over (FR-78).
    pub enum CallTransport {
        P2p => "p2p",
        Sfu => "sfu",
    }
}

string_enum! {
    /// Audio-only vs audio+video call (FR-79).
    pub enum CallKind {
        Audio => "audio",
        Video => "video",
    }
}

string_enum! {
    pub enum CallRoomState {
        Ringing => "ringing",
        Active => "active",
        Ended => "ended",
    }
}

string_enum! {
    pub enum CallInviteState {
        Ringing => "ringing",
        Accepted => "accepted",
        Declined => "declined",
        Cancelled => "cancelled",
    }
}

string_enum! {
    pub enum CallParticipantState {
        Joined => "joined",
        Left => "left",
    }
}

string_enum! {
    /// Coarse network-quality bucket (FR-82).
    pub enum CallNetworkQuality {
        Unknown => "unknown",
        Poor => "poor",
        Fair => "fair",
        Good => "good",
        Excellent => "excellent",
    }
}

string_enum! {
    /// Kinds of WebRTC signal relayed via the `WebRTCSignal` type.
    pub enum WebRTCSignalKind {
        Offer => "offer",
        Answer => "answer",
        Ice => "ice",
        Renegotiate => "renegotiate",
        SfuToken => "sfuToken",
        KeyEpoch => "keyEpoch",
    }
}

string_enum! {
    /// Media kind a producer/consumer carries.
    pub enum CallSfuMediaKind {
        Audio => "audio",
        Video => "video",
    }
}

string_enum! {
    /// Kinds of lightweight in-call data-channel envelope relayed via the
    /// `CallDataChannel` signal (AURA-316): ephemeral UX signaling that rides
    /// alongside the media plane rather than a `CallCommand` or a durable
    /// `CallEventStream` event.
    pub enum CallDataChannelKind {
        /// A transient emoji/reaction burst.
        Reaction => "reaction",
        /// A participant raising/lowering their hand.
        RaiseHand => "raiseHand",
        /// A live caption/transcription fragment.
        Caption => "caption",
    }
}

/// Canonical signal type name for WebRTC relay (`WEBRTC_SIGNAL_TYPE`).
pub const WEBRTC_SIGNAL_TYPE: &str = "WebRTCSignal";

/// Canonical signal type name for the in-call data-channel relay (AURA-316):
/// reactions / raise-hand / captions. Gated on the same call-membership check
/// as [`WEBRTC_SIGNAL_TYPE`] (FR-284), but kept as a distinct signal name so a
/// client can subscribe to/rate-limit UX chatter independently of SDP/ICE.
pub const CALL_DATA_CHANNEL_TYPE: &str = "CallDataChannel";

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CallRoomRecord {
    pub id: String,
    pub conversation_id: String,
    pub state: CallRoomState,
    pub created_by: String,
    pub kind: CallKind,
    pub created_at: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub started_at: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub ended_at: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub media_session_id: Option<String>,
    /// Plain `string` in TS (not `CallTransport`) — kept that way.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub transport: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CallInviteRecord {
    pub id: String,
    pub call_id: String,
    pub invitee_user_id: String,
    pub status: CallInviteState,
    pub invited_by: String,
    pub invited_at: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub responded_at: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CallParticipantRecord {
    pub id: String,
    pub call_id: String,
    pub user_id: String,
    pub device_id: String,
    pub state: CallParticipantState,
    pub joined_at: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub left_at: Option<String>,
    pub mic_enabled: bool,
    pub camera_enabled: bool,
    pub screen_sharing: bool,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub speaking: Option<bool>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub network_quality: Option<CallNetworkQuality>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CallMediaGrant {
    pub call_id: String,
    pub media_session_id: String,
    pub user_id: String,
    pub device_id: String,
    pub token: String,
    pub expires_at: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub connection: Option<IndexMap<String, String>>,
}

/// Partial media-state mutation for `setMediaState` (FR-82).
#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CallMediaStatePatch {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub mic_enabled: Option<bool>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub camera_enabled: Option<bool>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub screen_sharing: Option<bool>,
}

/// Discriminated union of every call control-plane command
/// (`CallCommandOp`). On the wire each command is a map whose first key is
/// the `op` discriminator string.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "op", rename_all = "camelCase")]
pub enum CallCommandOp {
    #[serde(rename_all = "camelCase")]
    Create {
        conversation_id: String,
        /// Users to invite. Must be non-empty and must exclude the caller.
        invitee_user_ids: Vec<String>,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        kind: Option<CallKind>,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        region_hint: Option<String>,
    },
    #[serde(rename_all = "camelCase")]
    Join { call_id: String },
    #[serde(rename_all = "camelCase")]
    Accept { call_id: String },
    #[serde(rename_all = "camelCase")]
    Leave { call_id: String },
    #[serde(rename_all = "camelCase")]
    End { call_id: String },
    #[serde(rename_all = "camelCase")]
    SetMediaState {
        call_id: String,
        media: CallMediaStatePatch,
    },
    #[serde(rename_all = "camelCase")]
    SfuConnectTransport {
        call_id: String,
        token: String,
        transport_id: String,
        /// Opaque mediasoup DTLS parameters, forwarded uninterpreted.
        dtls_parameters: Value,
    },
    #[serde(rename_all = "camelCase")]
    SfuProduce {
        call_id: String,
        token: String,
        transport_id: String,
        kind: CallSfuMediaKind,
        /// Opaque mediasoup RTP parameters, forwarded uninterpreted.
        rtp_parameters: Value,
    },
    #[serde(rename_all = "camelCase")]
    SfuConsume {
        call_id: String,
        token: String,
        transport_id: String,
        producer_id: String,
        /// Opaque mediasoup RTP capabilities, forwarded uninterpreted.
        rtp_capabilities: Value,
    },
}

string_enum! {
    /// Names of the supported command operations (`CallCommandName`).
    pub enum CallCommandName {
        Create => "create",
        Join => "join",
        Accept => "accept",
        Leave => "leave",
        End => "end",
        SetMediaState => "setMediaState",
        SfuConnectTransport => "sfuConnectTransport",
        SfuProduce => "sfuProduce",
        SfuConsume => "sfuConsume",
    }
}

impl CallCommandOp {
    /// The discriminator string this command serializes under.
    #[must_use]
    pub fn name(&self) -> CallCommandName {
        match self {
            Self::Create { .. } => CallCommandName::Create,
            Self::Join { .. } => CallCommandName::Join,
            Self::Accept { .. } => CallCommandName::Accept,
            Self::Leave { .. } => CallCommandName::Leave,
            Self::End { .. } => CallCommandName::End,
            Self::SetMediaState { .. } => CallCommandName::SetMediaState,
            Self::SfuConnectTransport { .. } => CallCommandName::SfuConnectTransport,
            Self::SfuProduce { .. } => CallCommandName::SfuProduce,
            Self::SfuConsume { .. } => CallCommandName::SfuConsume,
        }
    }
}

/// A call command frame body (`CallCommandPayload`).
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CallCommandPayload {
    pub request_id: String,
    pub command: CallCommandOp,
}

/// Result of `sfuProduce` (`CallSfuProduceResult`).
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CallSfuProduceResult {
    pub producer_id: String,
    pub kind: CallSfuMediaKind,
}

/// Result of `sfuConsume` (`CallSfuConsumeResult`).
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CallSfuConsumeResult {
    pub consumer_id: String,
    pub producer_id: String,
    pub kind: CallSfuMediaKind,
    pub rtp_parameters: Value,
}

/// The server's reply to a [`CallCommandPayload`] — fields populated per
/// command, failures ride the ordinary `Nack` frame.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CallCommandResultPayload {
    pub request_id: String,
    pub op: CallCommandName,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub room: Option<CallRoomRecord>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub invites: Option<Vec<CallInviteRecord>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub participant: Option<CallParticipantRecord>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub media_grant: Option<CallMediaGrant>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub invite: Option<CallInviteRecord>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub producer: Option<CallSfuProduceResult>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub consumer: Option<CallSfuConsumeResult>,
}
